source/common/config/utility.cc

```diff
@@ -81,20 +81,27 @@ void Utility::checkFilesystemSubscriptionBackingPath(const std::string& path) {
 void Utility::checkApiConfigSourceSubscriptionBackingCluster(
     const Upstream::ClusterManager::ClusterInfoMap& clusters,
     const envoy::api::v2::core::ApiConfigSource& api_config_source) {
-  if (api_config_source.cluster_names().size() != 1) {
-    // TODO(htuch): Add support for multiple clusters, #1170.
-    throw EnvoyException(
-        "envoy::api::v2::core::ConfigSource must have a singleton cluster name specified");
-  }
+  if (api_config_source.api_type() == envoy::api::v2::core::ApiConfigSource::GRPC) {
+    if (api_config_source.cluster_names().size() != 0) {
+      throw EnvoyException(
+          "envoy::api::v2::core::ConfigSource::GRPC must not have a cluster name specified");
+    }
+  } else {
+    if (api_config_source.cluster_names().size() != 1) {
+      // TODO(htuch): Add support for multiple clusters, #1170.
+      throw EnvoyException(
+          "envoy::api::v2::core::ConfigSource must have a singleton cluster name specified");
+    }
 
-  const auto& cluster_name = api_config_source.cluster_names()[0];
-  const auto& it = clusters.find(cluster_name);
-  if (it == clusters.end() || it->second.get().info()->addedViaApi() ||
-      it->second.get().info()->type() == envoy::api::v2::Cluster::EDS) {
-    throw EnvoyException(fmt::format(
-        "envoy::api::v2::core::ConfigSource must have a statically "
-        "defined non-EDS cluster: '{}' does not exist, was added via api, or is an EDS cluster",
-        cluster_name));
+    const auto& cluster_name = api_config_source.cluster_names()[0];
```

Comment:
I see what you meant here. You should have a condition here; if it's `api_type() == GRPC && api_config_soures.grpc_service()[0].has_envoy_grpc()` then you can verify `api_Config.grpc_Services()[0].envou_grpc().cluster_name()`.